File: internal/usthing/usthing.go
package usthing

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

// TokenManager 管理 Azure AD token 的获取与刷新
type TokenManager struct {
	mu          sync.Mutex
	accessToken string
	expiresAt   time.Time
	username    string
	password    string
}

var defaultTokenManager *TokenManager

func init() {
	defaultTokenManager = &TokenManager{
		username: os.Getenv("USTHING_USERNAME"),
		password: os.Getenv("USTHING_PASSWORD"),
	}
}

// SetCredentials 设置认证凭据
func SetCredentials(username, password string) {
	defaultTokenManager.mu.Lock()
	defer defaultTokenManager.mu.Unlock()
	defaultTokenManager.username = username
	defaultTokenManager.password = password
	defaultTokenManager.accessToken = "" // 强制重新获取
}

// ForceRefresh 强制刷新 token（当 API 返回 401 时调用）
func ForceRefresh() {
	defaultTokenManager.mu.Lock()
	defer defaultTokenManager.mu.Unlock()
	defaultTokenManager.accessToken = ""
	log.Println("[Auth] Token cache cleared — will re-acquire on next request")
}

func (tm *TokenManager) getAccessToken() (string, error) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	if tm.accessToken != "" && time.Now().Add(5*time.Minute).Before(tm.expiresAt) {
		return tm.accessToken, nil
	}

	if tm.username == "" || tm.password == "" {
		return "", fmt.Errorf("USTHING_USERNAME or USTHING_PASSWORD not set")
	}

	data := url.Values{}
	data.Set("grant_type", "password")
	data.Set("client_id", "04b07795-8ddb-461a-bbee-02f9e1bf7b46")
	data.Set("scope", "openid profile email offline_access")
	data.Set("username", tm.username)
	data.Set("password", tm.password)

	req, err := http.NewRequest("POST",
		"https://login.microsoftonline.com/c917f3e2-9322-4926-9bb3-daca730413ca/oauth2/v2.0/token",
		strings.NewReader(data.Encode()))
	if err != nil {
		return "", fmt.Errorf("error creating token request: %v", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("error fetching token: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("error reading token response: %v", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("token endpoint returned %d: %s", resp.StatusCode, string(body))
	}

	var tokenResp struct {
		AccessToken  string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		ExpiresIn    int    `json:"expires_in"`
	}
	if err := json.Unmarshal(body, &tokenResp); err != nil {
		return "", fmt.Errorf("error parsing token response: %v", err)
	}

	tm.accessToken = tokenResp.AccessToken
	tm.expiresAt = time.Now().Add(time.Duration(tokenResp.ExpiresIn) * time.Second)
	log.Printf("[Auth] Token refreshed, expires in %ds", tokenResp.ExpiresIn)

	return tm.accessToken, nil
}

// GetAccessToken 获取当前有效的 access token
func GetAccessToken() (string, error) {
	return defaultTokenManager.getAccessToken()
}

// generateHeaders 生成 HTTP 请求头
func generateHeaders() (http.Header, error) {
	token, err := defaultTokenManager.getAccessToken()
	if err != nil {
		return nil, err
	}

	headers := http.Header{}
	headers.Set("Accept", "application/json")
	headers.Set("Connection", "keep-alive")
	headers.Set("Cookie", "language=en-US")
	headers.Set("User-Agent", "USThing/428 CFNetwork/3860.100.1 Darwin/25.0.0")
	headers.Set("Authorization", "Bearer "+token)
	return headers, nil
}

// isAuthError 判断响应是否为 token 过期
func isAuthError(statusCode int, body []byte) bool {
	if statusCode == http.StatusUnauthorized {
		return true
	}
	bodyStr := string(body)
	return strings.Contains(bodyStr, "jwt malformed") ||
		strings.Contains(bodyStr, "JsonWebTokenError") ||
		strings.Contains(bodyStr, "Missing Authorization Header")
}

// doWithAuthRetry 执行 HTTP 请求，若遇到 401 则自动刷新 token 并重试一次
func doWithAuthRetry(req *http.Request) (*http.Response, error) {
	client := &http.Client{Timeout: 15 * time.Second}

	// 第一次尝试
	headers, err := generateHeaders()
	if err != nil {
		return nil, err
	}
	req.Header = headers

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}

	// 如果不是 401，直接返回
	body, _ := io.ReadAll(resp.Body)
	resp.Body.Close()

	if !isAuthError(resp.StatusCode, body) {
		// 重建 body reader 以便调用方读取
		resp.Body = io.NopCloser(strings.NewReader(string(body)))
		return resp, nil
	}

	log.Printf("[Auth] Received %d from API, refreshing token and retrying...", resp.StatusCode)

	// 强制刷新 token
	ForceRefresh()

	// 重试
	headers, err = generateHeaders()
	if err != nil {
		return nil, err
	}
	req.Header = headers

	resp2, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("retry after token refresh failed: %v", err)
	}

	log.Printf("[Auth] Retry completed with status %d", resp2.StatusCode)
	return resp2, nil
}

// ---- 数据结构 ----

// USThingTimeSlot 场地时间段
type USThingTimeSlot struct {
	FacilityID     int    `json:"facilityID"`
	TimeslotDate   string `json:"timeslotDate"`
	StartTime      string `json:"startTime"`
	EndTime        string `json:"endTime"`
	TimeslotStatus string `json:"timeslotStatus"`
	ActivityName   string `json:"activityName"`
}

// USThingTimeslotResponse 时段查询响应
type USThingTimeslotResponse struct {
	Status     string            `json:"status"`
	Message    string            `json:"message"`
	ErrorCode  string            `json:"errorCode"`
	FacilityID int               `json:"facilityID"`
	UserType   string            `json:"userType"`
	UstID      string            `json:"ustID"`
	StartDate  string            `json:"startDate"`
	EndDate    string            `json:"endDate"`
	TimeSlots  []USThingTimeSlot `json:"timeslot"`
}

// USThingBookingResponse 预订响应
type USThingBookingResponse struct {
	Status        string        `json:"status"`
	Message       string        `json:"message"`
	ErrorCode     string        `json:"errorCode"`
	TotalRecord   int           `json:"totalRecord"`
	UserType      string        `json:"userType"`
	UstID         string        `json:"ustID"`
	EmailAddr     string        `json:"emailAddr"`
	FacilityID    int           `json:"facilityID"`
	TimeslotDate  string        `json:"timeslotDate"`
	StartTime     string        `json:"startTime"`
	EndTime       string        `json:"endTime"`
	BookingRef    int           `json:"bookingRef"`
	CancelInd     *string       `json:"cancelInd"`
	BookingResult []interface{} `json:"bookingResult"`
}

// USThingFacility 设施信息
type USThingFacility struct {
	FacilityID   int    `json:"facilityID"`
	FacilityName string `json:"facilityName"`
	Location     string `json:"location"`
}

// USThingFacilityResponse 设施列表响应
type USThingFacilityResponse struct {
	Status      string            `json:"status"`
	Message     string            `json:"message"`
	TotalRecord int               `json:"totalRecord"`
	UserType    string            `json:"userType"`
	UstID       string            `json:"ustID"`
	Facilities  []USThingFacility `json:"facility"`
}

// USThingBookingInfo 预订信息
type USThingBookingInfo struct {
	FacilityID   int    `json:"facilityID"`
	FacilityName string `json:"facilityName"`
	Location     string `json:"location"`
	TimeslotDate string `json:"timeslotDate"`
	StartTime    string `json:"startTime"`
	EndTime      string `json:"endTime"`
	BookingRef   int    `json:"bookingRef"`
}

// USThingBookingInfoResponse 预订列表响应
type USThingBookingInfoResponse struct {
	Status      string              `json:"status"`
	Message     string              `json:"message"`
	ErrorCode   string              `json:"errorCode"`
	TotalRecord int                 `json:"totalRecord"`
	UserType    string              `json:"userType"`
	UstID       string              `json:"ustID"`
	EmailAddr   string              `json:"emailAddr"`
	Bookings    []USThingBookingInfo `json:"booking"`
}

const baseURL = "https://ms.api.usthing.xyz"

// ---- API 函数（全部使用 doWithAuthRetry 自动刷新） ----

// GetFacilities 获取所有设施列表 (v3)
func GetFacilities() (*USThingFacilityResponse, error) {
	req, err := http.NewRequest("GET", baseURL+"/v3/msapi/fbs/facilities", nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}
	log.Printf("[Facilities] GET %s", req.URL.String())

	resp, err := doWithAuthRetry(req)
	if err != nil {
		return nil, fmt.Errorf("error sending request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response: %v", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(body))
	}

	var response USThingFacilityResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	log.Printf("[Facilities] Found %d facilities", response.TotalRecord)
	return &response, nil
}

// GetAvailableTimeSlots 获取可用时间段 (v3)
func GetAvailableTimeSlots(ustID, userType, facilityID, startDate, endDate string) ([]USThingTimeSlot, error) {
	if ustID == "" {
		ustID = os.Getenv("USTHING_UST_ID")
	}
	if userType == "" {
		userType = "01"
	}

	url := fmt.Sprintf("%s/v3/msapi/fbs/facilityTimeslot?ustID=%s&userType=%s&facilityID=%s&startDate=%s&endDate=%s",
		baseURL, ustID, userType, facilityID, startDate, endDate)
	log.Printf("[Timeslot] GET %s", url)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}

	resp, err := doWithAuthRetry(req)
	if err != nil {
		return nil, fmt.Errorf("error sending request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response: %v", err)
	}

	log.Printf("[Timeslot] HTTP %d", resp.StatusCode)

	var response USThingTimeslotResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	if response.ErrorCode == "03" {
		log.Printf("[Timeslot] System closed: %s", response.Message)
		return nil, fmt.Errorf("system closed: %s", response.Message)
	}

	if response.Status != "200" {
		return nil, fmt.Errorf("unexpected status: %s, message: %s", response.Status, response.Message)
	}

	log.Printf("[Timeslot] Got %d timeslots", len(response.TimeSlots))

	availableSlots := make([]USThingTimeSlot, 0)
	for _, slot := range response.TimeSlots {
		if slot.TimeslotStatus == "Available" {
			availableSlots = append(availableSlots, slot)
		}
	}

	log.Printf("[Timeslot] Available: %d / %d", len(availableSlots), len(response.TimeSlots))
	return availableSlots, nil
}

// GetBookingInfo 获取当前预订列表 (v3)
func GetBookingInfo(ustID, userType string) (*USThingBookingInfoResponse, error) {
	if ustID == "" {
		ustID = os.Getenv("USTHING_UST_ID")
	}
	if userType == "" {
		userType = "01"
	}

	url := fmt.Sprintf("%s/v3/msapi/fbs/bookingInfo?ustID=%s&userType=%s", baseURL, ustID, userType)
	log.Printf("[BookingInfo] GET %s", url)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}

	resp, err := doWithAuthRetry(req)
	if err != nil {
		return nil, fmt.Errorf("error sending request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response: %v", err)
	}

	var response USThingBookingInfoResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("error parsing response: %v", err)
	}

	log.Printf("[BookingInfo] status=%s, totalRecord=%d", response.Status, response.TotalRecord)
	return &response, nil
}

// Booking 预订/取消场地 (v2)
func Booking(ustID, userType, facilityID, timeslotDate, startTime, endTime, cancelInd string) (*USThingBookingResponse, error) {
	if ustID == "" {
		ustID = os.Getenv("USTHING_UST_ID")
	}

	url := fmt.Sprintf("%s/v2/fbs/book?ustID=%s&userType=%s&facilityID=%s&timeslotDate=%s&startTime=%s&endTime=%s&cancelInd=%s",
		baseURL, ustID, userType, facilityID, timeslotDate, startTime, endTime, cancelInd)
	log.Printf("[Booking] POST %s", url)

	req, err := http.NewRequest("POST", url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %v", err)
	}

	resp, err := doWithAuthRetry(req)
	if err != nil {
		return nil, fmt.Errorf("error sending request: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response body: %v", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("received non-OK HTTP status: %s, body: %s", resp.Status, string(body))
	}

	var bookingResponse USThingBookingResponse
	if err := json.Unmarshal(body, &bookingResponse); err != nil {
		return nil, fmt.Errorf("error parsing JSON response: %v", err)
	}

	if bookingResponse.Status == "200" {
		log.Printf("[Booking] Success! bookingRef=%d", bookingResponse.BookingRef)
	} else {
		log.Printf("[Booking] Failed: %s (errorCode=%s)", bookingResponse.Message, bookingResponse.ErrorCode)
	}

	return &bookingResponse, nil
}

File: ts/sources/usthing.ts
import { USThingTimeSlot, USThingFacility, USThingBookingInfo } from "../types";

type HeaderMap = Record<string, string>;

const BASE_HEADERS: HeaderMap = {
  Accept: "application/json",
  Connection: "keep-alive",
  Cookie: "language=en-US",
  "User-Agent": "USThing/428 CFNetwork/3860.100.1 Darwin/25.0.0",
};

function buildHeaders(bearer?: string): HeaderMap {
  const headers: HeaderMap = { ...BASE_HEADERS };
  if (bearer) {
    headers.Authorization = bearer.startsWith("Bearer ")
      ? bearer
      : `Bearer ${bearer}`;
  }
  return headers;
}

// ---- Azure AD Token ----

interface AzureTokenResponse {
  token_type: string;
  scope: string;
  expires_in: number;
  access_token: string;
  refresh_token: string;
}

let cachedToken: { accessToken: string; expiresAt: number } | null = null;

let cachedCredentials: { username: string; password: string } | null = null;

/** 设置用于自动刷新的凭据 */
export function setCredentials(username: string, password: string): void {
  cachedCredentials = { username, password };
}

/**
 * 通过 Azure AD ROPC 获取 access token。
 * 若 force=true 则跳过缓存强制刷新。
 */
export async function acquireToken(
  username: string,
  password: string,
  options: { fetchImpl?: typeof fetch; force?: boolean } = {}
): Promise<string> {
  if (!options.force && cachedToken && Date.now() + 5 * 60 * 1000 < cachedToken.expiresAt) {
    return cachedToken.accessToken;
  }

  const { fetchImpl = fetch } = options;
  const body = new URLSearchParams();
  body.set("grant_type", "password");
  body.set("client_id", "04b07795-8ddb-461a-bbee-02f9e1bf7b46");
  body.set("scope", "openid profile email offline_access");
  body.set("username", username);
  body.set("password", password);

  const response = await fetchImpl(
    "https://login.microsoftonline.com/c917f3e2-9322-4926-9bb3-daca730413ca/oauth2/v2.0/token",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        Accept: "application/json",
      },
      body: body.toString(),
    }
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(
      `Azure AD token request failed with ${response.status}: ${text}`
    );
  }

  const data = (await response.json()) as AzureTokenResponse;
  cachedToken = {
    accessToken: data.access_token,
    expiresAt: Date.now() + data.expires_in * 1000,
  };
  cachedCredentials = { username, password };

  console.log(`[Auth] Token refreshed, expires in ${data.expires_in}s`);
  return data.access_token;
}

/** 清除缓存 token（强制下次请求重新获取） */
export function clearTokenCache(): void {
  cachedToken = null;
}

/**
 * 强制刷新 token — 当 API 返回 401 时调用。
 * 模拟 App 行为：清除过期 token，用已缓存的凭据重新获取。
 */
export async function forceRefreshToken(
  options: { fetchImpl?: typeof fetch } = {}
): Promise<string> {
  clearTokenCache();
  if (!cachedCredentials) {
    throw new Error(
      "Cannot refresh token: no cached credentials. Call setCredentials() or acquireToken() first."
    );
  }
  return acquireToken(cachedCredentials.username, cachedCredentials.password, {
    ...options,
    force: true,
  });
}

// ---- Auth retry helper ----

function isAuthError(status: number, body: string): boolean {
  if (status === 401) return true;
  const lower = body.toLowerCase();
  return (
    lower.includes("jwt malformed") ||
    lower.includes("jsonwebtokenerror") ||
    lower.includes("missing authorization header")
  );
}

/**
 * fetchWithAuthRetry 包装 fetch，检测 401 → 自动刷新 token → 重试一次。
 * 完全模仿 App 的 token 过期自动刷新行为。
 */
async function fetchWithAuthRetry(
  url: string,
  init: RequestInit,
  bearer?: string,
  fetchImpl: typeof fetch = fetch
): Promise<Response> {
  // 动态注入最新 token
  const makeHeaders = () => {
    // 每次都重新 build headers，确保 token 是最新的
    return buildHeaders(bearer);
  };

  const req: RequestInit = { ...init, headers: makeHeaders() };

  let response = await fetchImpl(url, req);

  if (!isAuthError(response.status, "")) {
    return response;
  }

  // 401 — 读取 body 二次确认
  const clonedBody = await response.clone().text();
  if (!isAuthError(response.status, clonedBody)) {
    return response;
  }

  console.log(
    `[Auth] Received ${response.status} from API, refreshing token and retrying...`
  );

  try {
    bearer = await forceRefreshToken({ fetchImpl });
  } catch (e) {
    console.error("[Auth] Token refresh failed:", e);
    return response; // 返回原始 401，让调用方处理
  }

  // 重试
  console.log("[Auth] Retrying request with refreshed token...");
  req.headers = makeHeaders();
  response = await fetchImpl(url, req);
  console.log(`[Auth] Retry completed with status ${response.status}`);
  return response;
}

// ---- Response types (internal) ----

interface USThingTimeslotResponse {
  status: string;
  message: string;
  errorCode?: string;
  timeslot: USThingTimeSlot[];
}

interface USThingBookingResponse {
  status: string;
  message: string;
  errorCode: string;
  totalRecord: number;
  userType: string;
  ustID: string;
  emailAddr: string;
  facilityID: number;
  timeslotDate: string;
  startTime: string;
  endTime: string;
  bookingRef: number;
  cancelInd: string | null;
  bookingResult: unknown[];
}

interface USThingFacilityResponse {
  status: string;
  message: string;
  totalRecord: number;
  facility: USThingFacility[];
}

interface USThingBookingInfoResponse {
  status: string;
  message: string;
  errorCode: string;
  totalRecord: number;
  booking: USThingBookingInfo[];
}

// ---- API Functions (all use fetchWithAuthRetry) ----

/** 获取所有设施列表 (v3) */
export async function getFacilities(
  options: { fetchImpl?: typeof fetch; bearer?: string } = {}
): Promise<USThingFacility[]> {
  const { fetchImpl = fetch, bearer } = options;
  const url = "https://ms.api.usthing.xyz/v3/msapi/fbs/facilities";
  console.log("[USThing] GET facilities");

  const response = await fetchWithAuthRetry(
    url,
    { method: "GET" },
    bearer,
    fetchImpl
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`USThing facilities failed with ${response.status}: ${text}`);
  }

  const data = (await response.json()) as USThingFacilityResponse;
  console.log(`[USThing] Facilities: ${data.totalRecord} found`);
  return data.facility ?? [];
}

/** 获取可用时间段 (v3) */
export async function getAvailableTimeSlots(
  ustID: string,
  userType: string,
  facilityID: string,
  startDate: string,
  endDate: string,
  options: { fetchImpl?: typeof fetch; bearer?: string } = {}
): Promise<USThingTimeSlot[]> {
  const { fetchImpl = fetch, bearer } = options;
  const url = `https://ms.api.usthing.xyz/v3/msapi/fbs/facilityTimeslot?ustID=${encodeURIComponent(
    ustID
  )}&userType=${encodeURIComponent(
    userType
  )}&facilityID=${encodeURIComponent(
    facilityID
  )}&startDate=${encodeURIComponent(
    startDate
  )}&endDate=${encodeURIComponent(endDate)}`;

  console.log(
    `[USThing] GET facilityTimeslot facility=${facilityID} range=${startDate}→${endDate}`
  );

  const response = await fetchWithAuthRetry(
    url,
    { method: "GET" },
    bearer,
    fetchImpl
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`USThing timeslot failed with ${response.status}: ${text}`);
  }

  const data = (await response.json()) as USThingTimeslotResponse;

  if (data.errorCode === "03") {
    throw new Error(`USThing system closed: ${data.message}`);
  }

  if (data.status !== "200") {
    throw new Error(`Unexpected status: ${data.status} - ${data.message}`);
  }

  return data.timeslot ?? [];
}

/** 获取当前预订列表 (v3) */
export async function getBookingInfo(
  ustID: string,
  userType: string,
  options: { fetchImpl?: typeof fetch; bearer?: string } = {}
): Promise<USThingBookingInfo[]> {
  const { fetchImpl = fetch, bearer } = options;
  const url = `https://ms.api.usthing.xyz/v3/msapi/fbs/bookingInfo?ustID=${encodeURIComponent(
    ustID
  )}&userType=${encodeURIComponent(userType)}`;

  console.log("[USThing] GET bookingInfo");

  const response = await fetchWithAuthRetry(
    url,
    { method: "GET" },
    bearer,
    fetchImpl
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`USThing bookingInfo failed with ${response.status}: ${text}`);
  }

  const data = (await response.json()) as USThingBookingInfoResponse;
  console.log(`[USThing] Bookings: ${data.totalRecord}`);
  return data.booking ?? [];
}

/** 预订/取消场地 (v2 — 此端点仍有效) */
export async function booking(
  ustID: string,
  userType: string,
  facilityID: string,
  timeslotDate: string,
  startTime: string,
  endTime: string,
  cancelInd: string,
  options: { fetchImpl?: typeof fetch; bearer?: string } = {}
): Promise<USThingBookingResponse> {
  const { fetchImpl = fetch, bearer } = options;
  const url = `https://ms.api.usthing.xyz/v2/fbs/book?ustID=${encodeURIComponent(
    ustID
  )}&userType=${encodeURIComponent(
    userType
  )}&facilityID=${encodeURIComponent(
    facilityID
  )}&timeslotDate=${encodeURIComponent(
    timeslotDate
  )}&startTime=${encodeURIComponent(
    startTime
  )}&endTime=${encodeURIComponent(
    endTime
  )}&cancelInd=${encodeURIComponent(cancelInd)}`;

  console.log(
    `[USThing] POST book facility=${facilityID} ${timeslotDate} ${startTime}-${endTime} cancelInd=${cancelInd}`
  );

  const response = await fetchWithAuthRetry(
    url,
    { method: "POST" },
    bearer,
    fetchImpl
  );

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`USThing booking failed with ${response.status}: ${text}`);
  }

  const data = (await response.json()) as USThingBookingResponse;

  if (data.status !== "200") {
    throw new Error(`USThing booking unsuccessful: ${data.message}`);
  }

  console.log(
    `[USThing] Booking ${
      cancelInd === "Y" ? "cancelled" : "created"
    }, bookingRef=${data.bookingRef}`
  );
  return data;
}
